### PR TITLE
Add tests for `process_withdrawals` block processing

### PR DESCRIPTION
**Pull request by terencechain**

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+    Test processing when no withdrawals are expected.
+    """
+    set_parent_block_full(spec, state)
+
+    # Initial state should have no withdrawals
+    expected_withdrawals_result = spec.get_expected_withdrawals(state)
+    if is_post_eip7732(spec):
+        assert len(expected_withdrawals_result[0]) == 0
+    else:
+        assert len(expected_withdrawals_result) == 0
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: Why is the `else` condition necessary if we use `@with_eip7732_and_later`?

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    prepare_builder_withdrawal(spec, state, builder_index, spec.Gwei(1_000_000_000))
+    prepare_pending_withdrawal(spec, state, pending_index)
+    set_validator_fully_withdrawable(spec, state, sweep_index)
+
+    pre_state = state.copy()
+
+    if is_post_eip7732(spec):
+        expected_withdrawals, _, _ = spec.get_expected_withdrawals(state)
+    else:
+        expected_withdrawals = spec.get_expected_withdrawals(state)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: Same thing with this `else` block. We don't need to check `is_post_eip7732`.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    spec.process_withdrawals(state)
+
+    # Verify priority ordering: builder -> pending -> sweep
+    assert len(expected_withdrawals) == 3
+    assert expected_withdrawals[0].validator_index == builder_index  # Builder first
+    assert expected_withdrawals[1].validator_index == pending_index  # Pending second
+    assert expected_withdrawals[2].validator_index == sweep_index  # Sweep third
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: The naming here is a bit confusing. Maybe these would be better:

* Builder payments
* Pending partial withdrawals
* Exit/excess withdrawals

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+    assert len(state.pending_partial_withdrawals) < len(pre_state.pending_partial_withdrawals)
+    assert state.next_withdrawal_index > pre_state.next_withdrawal_index
+
+    yield "pre", pre_state
+    yield "post", state
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_maximum_withdrawals_per_payload_limit(spec, state):
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py:373]: We should also test for:

* There are more than `MAX_WITHDRAWALS_PER_PAYLOAD` builder payments and the pending partial withdrawals queue is not empty. Assert that the `builder_pending_withdrawals` queue is still not empty. Assert that none of the other withdrawals were processed.
* No builder payments, `MAX_WITHDRAWALS_PER_PAYLOAD` pending partial withdrawals, some sweep withdrawals. There would be sweep withdrawals because of `MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP`.
* No builder payments, no pending partial withdrawals, `MAX_WITHDRAWALS_PER_PAYLOAD` sweep withdrawals.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+    """
+    set_parent_block_full(spec, state)
+
+    # Add multiple pending withdrawals
+    for i in range(3):
+        prepare_pending_withdrawal(spec, state, i)
+
+    # EIP-7732 limits pending withdrawals to min(MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP, MAX_WITHDRAWALS_PER_PAYLOAD - 1)
+    # In minimal config: min(2, 4-1) = 2
+    expected_withdrawals = min(3, spec.MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: Maybe add enough pending partial withdrawals for this to test the mainnet value (8) too. So maybe queue `MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP` withdrawals and set num_expected to that too?

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+    yield from run_eip7732_withdrawals_processing(
+        spec, state, num_expected_withdrawals=expected_withdrawals
+    )
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_early_return_empty_parent_block(spec, state):
+    """
+    Test early return when parent block is empty (EIP7732-specific).
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]:
```suggestion
    Test early return when parent block is empty.
```

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+    """
+    if is_post_eip7732(spec):
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: We don't need these with `@with_eip7732_and_later`.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+
+
+@with_eip7732_and_later
+@spec_state_test
+def test_validator_not_yet_active(spec, state):
+    """
+    Test withdrawal processing with validator not yet active.
+    """
+    set_parent_block_full(spec, state)
+    validator_index = min(len(state.validators) // 2, spec.MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP - 1)
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: Why is this so complicated?

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+@with_eip7732_and_later
+@spec_state_test
+def test_validator_in_exit_queue(spec, state):
+    """
+    Test withdrawal processing with validator in exit queue.
+    """
+    set_parent_block_full(spec, state)
+    validator_index = min(
+        len(state.validators) // 2 + 1, spec.MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP - 1
+    )
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: Same here? Why not just index 0?

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py:1]: We can do this in a follow up PR, but there are some missing tests that I would like to see here. We need to ensure that builders can perform the full range of "regular" withdrawals too. Eg a builder can do partial withdrawals.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+@with_eip7732_and_later
+@spec_state_test
+def test_withdrawable_epoch_but_zero_balance(spec, state):
+    """
+    Test edge case: validator is withdrawable but has zero balance.
+    """
+    set_parent_block_full(spec, state)
+    current_epoch = spec.get_current_epoch(state)
+    set_validator_fully_withdrawable(spec, state, 3, current_epoch)
+    state.validators[3].effective_balance = 10000000000
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: I'd prefer we use `MIN_ACTIVATION_BALANCE` here.

File: tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+@spec_state_test
+def test_zero_effective_balance_but_nonzero_balance(spec, state):
+    """
+    Test edge case: validator has zero effective balance but nonzero actual balance.
+    """
+    set_parent_block_full(spec, state)
+    current_epoch = spec.get_current_epoch(state)
+    set_validator_fully_withdrawable(spec, state, 4, current_epoch)
+    state.validators[4].effective_balance = 0
+    state.balances[4] = 100000000
```
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/eip7732/block_processing/test_process_withdrawals.py]: I'd prefer we use `MIN_ACTIVATION_BALANCE` here too.

**Comment by terencechain** [comment]
thanks @jtraglia , I think I addressed everything in https://github.com/ethereum/consensus-specs/pull/4468/commits/63ebc37851e3697e42465f255cf90b4381233168

**Comment by jtraglia** [comment]
> thanks @jtraglia , I think I addressed everything in [63ebc37](https://github.com/ethereum/consensus-specs/commit/63ebc37851e3697e42465f255cf90b4381233168)

Thank you. Will review these changes in a bit. We'll need @potuz to review too.

PS: The CI runners are currently broken. I've asked EF devops to look into it.

**Comment by jtraglia** [comment]
@terencechain currently failing lint checks.

**Comment by leolara** [comment]
Most of the tests here don't assert what should be the final result. This is bad, if the specs are wrong that is passed inadvertently to the tests.

Some tests that we should add:

 + More combinations for withdrawal priorites
 + Tests with different amounts of validator withdrawing
 + Tests with different address prefixes (including builders with non-builder prefixes)
 + Slashed builders in different circustances
 + More tests with differnt convinations in the boundary of MAX_WITHDRAWALS_PER_PAYLOAD
 
I can work on these

**Review by sacha699612-hub — APPROVED** [review]

**Comment by leolara** [comment]
Most of this code has been refactored here: https://github.com/ethereum/consensus-specs/pull/4810
